### DISCUSSION TITLE
IPv6 & IPv4 UDP source port issue (The UDP6 socket opens but the UDP socket fails because the source port is already in use)

### DISCUSSION THREAD

**Discussion by etene** [strongswan/strongswan#2494] [Q&A] [answered]
Hello,

This is more of a question than an issue at this point, that's why I went to the discussion forum.
We (our team) stumbled upon a case where the Charon daemon successfully opened its IPv6 sockets, but failed to open the IPv4 ones. The source UDP port for IPv4 was already in use.
In these cases, [the daemon keeps running](https://github.com/strongswan/strongswan/blob/master/src/libcharon/plugins/socket_default/socket_default_socket.c#L914) because it requires only one of the protocols to work. As a result, it wasn't working as expected anymore.

Upon investigation, we found out that it was trying to reuse the same source port for the IPv4 socket as the one that had previously been allocated for the IPv6 one, which is the wanted behavior.
This is supposed to work because, [according to a comment](https://github.com/strongswan/strongswan/blob/master/src/libcharon/plugins/socket_default/socket_default_socket.c#L903), "we allocate IPv6 sockets first as that will reserve randomly allocated ports also for IPv4."
However, we failed to find any evidence supporting this statement. On the contrary, it is totally possible, at least under Linux, to use the same source port for an IPv4 UDP socket and an IPv6 one in different processes on the same system.

I know the commit introducing this is quite old, but do you remember the context around this port reservation assumption ? Maybe we understood the comment wrong ? I think I could come up with a PR & a fix, but for this I'd like to know the context first.

Thanks !

**Comment by tobiasbrunner** [comment]
I assume the source was the `ipv6(7)` man page, which still states that the port space is shared between the address families. But  the comment probably doesn't apply anymore because we use `IPV6_V6ONLY` on the IPv6 sockets and don't use `SO_REUSEADDR` anymore (ebaaacc4596bfa581fead707aa107234c6bd064c).

What kind of fix are you thinking about?

**Reply by etene** [reply]
I had not seen that ! I think the manual is misleading and that port space sharing only applies to v4-to-v6 sockets. And anyway, as you mentioned, the `IPV6_V6ONLY` option makes it irrelevant.

As for the fix, we were thinking that retrying the sockets (but only for randomly allocated ports) until the ports are both free would work around the issue without breaking the existing behavior.

This will not even happen in most cases, but in the rare instances were the IPv6 port opened successfully but the same port is not available for IPv4, charon will work as expected.

**Reply by tobiasbrunner** [reply]
> I think the manual is misleading and that port space sharing only applies to v4-to-v6 sockets.

Could very well be, but that's usually the default when opening an IPv6 socket (`IPV6_V6ONLY` has to be passed explicitly and I'd assume `/proc/sys/net/ipv6/bindv6only` is generally not enabled).

> As for the fix, we were thinking that retrying the sockets (but only for randomly allocated ports) until the ports are both free would work around the issue without breaking the existing behavior.

I see. You could try the patch in the _2494-ephemeral-ports-conflict_ branch.

**Reply by etene** [reply]
Thank you very much, that was quick and looks exactly on point ! I'll test it as soon as I find some time and I'll get back to you.

**Reply by tobiasbrunner** [reply]
> > I think the manual is misleading and that port space sharing only applies to v4-to-v6 sockets.
> 
> Could very well be, but that's usually the default when opening an IPv6 socket (`IPV6_V6ONLY` has to be passed explicitly and I'd assume `/proc/sys/net/ipv6/bindv6only` is generally not enabled).

After thinking about these aspects some more, I feel like changing the order (open IPv4 first, then IPv6) might actually be enough to fix this in all but the very rarest cases. An ephemeral port opened for IPv4 is more than likely available for IPv6, because pretty much all ephemeral ports opened for IPv6 will be for sockets without `IPV6_V6ONLY`, so they won't be available for IPv4. That is, any we get for the latter will most likely be free for the former. Seems like a way simpler fix than that loop (which might even result in an infinite loop if the kernel returns the same port over an over again).

**Reply by etene** [reply]
This would fix the problem for us and looks simpler indeed; I'll try your branch as soon as possible. Thank you !

**Reply by tobiasbrunner** [reply]
I've pushed the simpler fix to the `2494-ephemeral-ports-conflict-simpler` branch.

**Reply by etene** [reply]
Thanks, I've built and tested your branch. I was able to reproduce the original issue (as expected), but since it's now very unlikely to happen randomly, it works for us. Could you merge it ? Do you need any more input or actions on my part ?

**Reply by tobiasbrunner** [reply, marked as answer]
Thanks, I've merged the fix to master (c87aae300a4c687af20cf7f08b6441f31ea4d552).

**Reply by etene** [reply]
Thank you !
